quickcheck-dynamic/src/Test/QuickCheck/StateModel.hs

```diff
@@ -231,110 +304,256 @@ instance Semigroup (Actions state) where
-              )
-            ]
-     where
-      satisfyPrecondition = sized $ \n -> go n (2 * n) [] -- idea copied from suchThatMaybe
-      go m n rej
-        | m > n = return (Nothing, rej)
-        | otherwise = do
-            a <- resize m $ arbitraryAction s
-            case a of
-              Some act ->
-                if precondition s act
-                  then return (Just (Some act), rej)
-                  else go (m + 1) n (actionName act : rej)
-              Error _ ->
-                go (m + 1) n rej
+    where
+      arbActions :: Annotated state -> Int -> Gen ([Step state], [String])
+      arbActions s step = sized $ \n ->
+        let w = n `div` 2 + 1
+         in frequency
+              [ (1, return ([], []))
+              ,
+                ( w
+                , do
+                    (mact, rej) <- satisfyPrecondition
+                    case mact of
+                      Just (Some act) -> do
+                        (as, rejected) <- arbActions (computeNextState s act (Var step)) (step + 1)
+                        return ((Var step := act) : as, rej ++ rejected)
+                      Nothing ->
+                        return ([], [])
+                )
+              ]
+        where
+          satisfyPrecondition = sized $ \n -> go n (2 * n) [] -- idea copied from suchThatMaybe
+          go m n rej
+            | m > n = return (Nothing, rej)
+            | otherwise = do
+                a <- resize m $ computeArbitraryAction s
+                case a of
+                  Some act ->
+                    if computePrecondition s act
+                      then return (Just (Some act), rej)
+                      else go (m + 1) n (actionName act : rej)
 
   shrink (Actions_ rs as) =
     map (Actions_ rs) (shrinkSmart (map (prune . map fst) . shrinkList shrinker . withStates) as)
-   where
-    shrinker (Var i := act, s) = [(Var i := act', s) | Some act' <- shrinkAction s act]
+    where
+      shrinker (Var i := act, s) = [(Var i := act', s) | Some act' <- computeShrinkAction s act]
+
+-- Running state models
+
+newtype VarContext = VarCtx (Set (Any Var)) deriving (Semigroup) via Set (Any Var)
+
+instance Show VarContext where
+  show (VarCtx vs) =
+    "[" ++ intercalate ", " (map showBinding . sortBy (comparing getIdx) $ Set.toList vs) ++ "]"
+    where
+      getIdx (Some (Var i)) = i
+      showBinding :: Any Var -> String
+      -- The use of typeRep here is on purpose to avoid printing `Var` unnecessarily.
+      showBinding (Some v) = show v ++ " :: " ++ show (typeRep v)
+
+data Annotated state = Metadata
+  { vars :: VarContext
+  , underlyingState :: state
+  }
+
+instance Show state => Show (Annotated state) where
+  show (Metadata ctx s) = show ctx ++ " |- " ++ show s
+
+isWellTyped :: Typeable a => Var a -> VarContext -> Bool
+isWellTyped v (VarCtx ctx) = Some v `Set.member` ctx
+
+emptyContext :: VarContext
+emptyContext = VarCtx mempty
+
+-- TODO: check invariants??
```

Comment:
No variable index is used twice at different types. This invariant should always hold unless you do something super crazy - so it might not be worth while actually checking it all the time (but on the other hand the context won't be big enough for the overhead to actually impact performance measurably).

Reply:
Suggestion: leave the TODO as-is but with a more precise description of the expected invariant to hold. Then someone can write a property later on to check the invariant holds :)